src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1027 @@
+    FIRST_SEQ_ID,
+};
+use crate::executor::prelude::*;
+use crate::executor::{
+    Barrier, BoxedMessageStream, Message, StreamExecutorError, StreamExecutorResult,
+};
+
+type StateStoreStream<S> = Pin<Box<LogStoreItemMergeStream<TimeoutAutoRebuildIter<S>>>>;
+type ReadFlushedChunkFuture = BoxFuture<'static, LogStoreResult<(ChunkId, StreamChunk, u64)>>;
+
+struct SyncedKvLogStoreExecutor<S: StateStore, LS: LocalStateStore> {
+    table_id: TableId,
+    read_metrics: KvLogStoreReadMetrics,
+    metrics: KvLogStoreMetrics,
+    serde: LogStoreRowSerde,
+    seq_id: SeqIdType,
+    truncation_offset: Option<ReaderTruncationOffsetType>,
+
+    // Upstream
+    upstream: Executor,
+
+    // Log store state
+    flushed_chunk_future: Option<ReadFlushedChunkFuture>,
+    state_store: S,
+    local_state_store: LS,
+    buffer: Mutex<SyncedLogStoreBuffer>,
+}
+// Stream interface
+impl<S: StateStore<Local = LS>, LS: LocalStateStore> SyncedKvLogStoreExecutor<S, LS> {
+    #[allow(clippy::too_many_arguments, dead_code)]
+    pub async fn new(
+        table_id: u32,
+        read_metrics: KvLogStoreReadMetrics,
+        metrics: KvLogStoreMetrics,
+        serde: LogStoreRowSerde,
+        seq_id: SeqIdType,
+        state_store: S,
+        buffer_max_size: usize,
+        upstream: Executor,
+    ) -> Self {
+        let local_state_store = state_store
+            .new_local(NewLocalOptions {
+                table_id: TableId { table_id },
+                op_consistency_level: OpConsistencyLevel::Inconsistent,
+                table_option: TableOption {
+                    retention_seconds: None,
+                },
+                is_replicated: false,
+                vnodes: serde.vnodes().clone(),
+            })
+            .await;
+        Self {
+            table_id: TableId::new(table_id),
+            read_metrics,
+            metrics: metrics.clone(),
+            serde,
+            seq_id,
+            truncation_offset: None,
+            flushed_chunk_future: None,
+            state_store,
+            local_state_store,
+            buffer: Mutex::new(SyncedLogStoreBuffer {
+                buffer: VecDeque::new(),
+                max_size: buffer_max_size,
+                next_chunk_id: 0,
+                metrics,
+            }),
+            upstream,
+        }
+    }
+}
+
+// Stream interface
+impl<S: StateStore, LS: LocalStateStore> SyncedKvLogStoreExecutor<S, LS> {
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub async fn execute_inner(mut self) {
+        let mut input = self.upstream.execute();
+        let Some(msg) = input.next().await else {
+            bail!("Expected a barrier message, got end of stream")
+        };
+        let barrier = match msg? {
+            Message::Barrier(barrier) => barrier,
+            other => bail!("Expected a barrier message, got {:?}", other),
+        };
+        let mut state_store_stream = Some(
```

Comment:
We should yield the barrier first and then init the local state store. Otherwise, we will get stuck at `LocalStateStore::init`.